### PR TITLE
Versioning store changes

### PR DESCRIPTION
**Pull request by chris124567**
Re: #201

Due to some interdependent things, it's difficult to get everything to build.  But if you run the sqlite tests specifically:

```
go test ./sia/persist/sqlite/ -v -count=1 
```

They all pass.

Before:
```
goos: linux
goarch: amd64
pkg: github.com/SiaFoundation/s3d/sia/persist/sqlite
cpu: Intel(R) Core(TM) i7-8665U CPU @ 1.90GHz
BenchmarkListObjects/no_delimiter_no_prefix-4         	     322	   3827107 ns/op	  377759 B/op	    8572 allocs/op
BenchmarkListObjects/slash/root_delimiter-4           	      20	  55469415 ns/op	 2221026 B/op	   53194 allocs/op
BenchmarkListObjects/slash/random_without_delimiter-4 	     656	   1709676 ns/op	  155647 B/op	    3462 allocs/op
BenchmarkListObjects/slash/random_with_delimiter-4    	    1510	    833290 ns/op	   33968 B/op	     724 allocs/op
BenchmarkListObjects/slash/folder_bottom_delimiter-4  	    6331	    188681 ns/op	    9998 B/op	     181 allocs/op
BenchmarkListObjects/slash/folder_delimiter-4         	    1552	    783232 ns/op	   33772 B/op	     724 allocs/op
BenchmarkListObjects/backslash/root_delimiter-4       	      19	  56824941 ns/op	 2220394 B/op	   53202 allocs/op
BenchmarkListObjects/backslash/random_without_delimiter-4         	     596	   1775787 ns/op	  155202 B/op	    3451 allocs/op
BenchmarkListObjects/backslash/random_with_delimiter-4            	    1214	    865385 ns/op	   33974 B/op	     724 allocs/op
BenchmarkListObjects/backslash/folder_bottom_delimiter-4          	    7107	    188810 ns/op	    9995 B/op	     181 allocs/op
BenchmarkListObjects/backslash/folder_delimiter-4                 	    1593	    769560 ns/op	   33770 B/op	     724 allocs/op
PASS
ok  	github.com/SiaFoundation/s3d/sia/persist/sqlite	87.042s
```

After:
```
goos: linux
goarch: amd64
pkg: github.com/SiaFoundation/s3d/sia/persist/sqlite
cpu: Intel(R) Core(TM) i7-8665U CPU @ 1.90GHz
BenchmarkListObjects/no_delimiter_no_prefix-4         	     296	   4011547 ns/op	  395252 B/op	    8550 allocs/op
BenchmarkListObjects/slash/root_delimiter-4           	      15	  72266987 ns/op	 2369141 B/op	   51211 allocs/op
BenchmarkListObjects/slash/random_without_delimiter-4 	     782	   1652186 ns/op	  142418 B/op	    3013 allocs/op
BenchmarkListObjects/slash/random_with_delimiter-4    	    1106	   1092842 ns/op	   36048 B/op	     702 allocs/op
BenchmarkListObjects/slash/folder_bottom_delimiter-4  	    5630	    231283 ns/op	   10322 B/op	     177 allocs/op
BenchmarkListObjects/slash/folder_delimiter-4         	    1182	   1037541 ns/op	   35840 B/op	     701 allocs/op
BenchmarkListObjects/backslash/root_delimiter-4       	      15	  70600722 ns/op	 2365412 B/op	   51205 allocs/op
BenchmarkListObjects/backslash/random_without_delimiter-4         	     608	   1803499 ns/op	  148958 B/op	    3154 allocs/op
BenchmarkListObjects/backslash/random_with_delimiter-4            	    1118	   1096305 ns/op	   36049 B/op	     702 allocs/op
BenchmarkListObjects/backslash/folder_bottom_delimiter-4          	    5480	    235659 ns/op	   10317 B/op	     177 allocs/op
BenchmarkListObjects/backslash/folder_delimiter-4                 	    1158	   1055111 ns/op	   35848 B/op	     701 allocs/op
PASS
ok  	github.com/SiaFoundation/s3d/sia/persist/sqlite	114.867s
```

### CODE REVIEW

File: sia/persist/sqlite/lifecycle.go [hunk 1]
```diff
@@ -184,21 +198,14 @@ func (s *Store) ExpireObjects(bucketID int64, prefix string, before time.Time, l
...
-				if err := insertOrphan(tx, *oldID); err != nil {
-					return err
-				}
-			}
-			orphanFile, orphanSize, err := newOrphanedFile(tx, filename, size)
-			if err != nil {
+			res, err := deleteCurrentObject(tx, bucketID, name, status, s3.ObjectID{Key: name})
+			if errors.Is(err, sql.ErrNoRows) {
+				continue // current version vanished concurrently; nothing to do
+			} else if err != nil {
```
[the remark this answers is not in the document]
Comment (chris124567) [reply sia/persist/sqlite/lifecycle.go:212]: We only have 1 writer

File: sia/persist/sqlite/store.go [hunk 2]
```diff
@@ -130,4 +132,6 @@ func OpenDatabase(fp string, log *zap.Logger) (*Store, error) {
 	return store, nil
 }
 
-var _ sia.Store = &Store{}
+// Disabled for the versioning-3-sqlite split: the sia.Store interface this
+// must satisfy is in the follow-up branch (versioning-3), which restores it.
+// var _ sia.Store = &Store{}
```
[the remark this answers is not in the document]
Comment (chris124567) [reply sia/persist/sqlite/store.go]: yes that's intentionaal see #214

File: sia/persist/sqlite/objects.go [hunk 3]
```diff
@@ -177,21 +247,26 @@ func getObject(tx *txn, obj *objects.Object, bid int64, name string, partNumber
...
-// is returned so the caller can remove it from disk and release its disk usage.
-func (s *Store) PutObject(accessKeyID, bucket, name string, contentMD5 [16]byte, meta map[string]string, length int64, fileName *string) (orphanFile string, orphanSize int64, _ error) {
+// PutObject stores the object and returns the wire-encoded version ID to
+// report ("" on a suspended or unversioned bucket, since neither reports a
+// version). An enabled bucket creates a new version; otherwise the null version
+// is overwritten, orphaning any prior object ID or pending file that is no
+// longer referenced (the latter returned so the caller can remove it from disk).
+func (s *Store) PutObject(accessKeyID, bucket, name string, contentMD5 [16]byte, meta map[string]string, length int64, fileName *string) (versionID string, orphan objects.OrphanedFile, _ error) {
 	err := s.transaction(func(tx *txn) error {
 		bid, err := bucketID(tx, accessKeyID, bucket)
```
[the remark this answers is not in the document]
Comment (chris124567) [reply sia/persist/sqlite/objects.go]: fixed

File: sia/persist/sqlite/buckets_test.go [hunk 4]
```diff
@@ -0,0 +1,52 @@
+package sqlite
```
Comment (peterjan) [inline sia/persist/sqlite/buckets_test.go:1]: perhaps buckets_test.go would be a better file name

File: sia/persist/sqlite/buckets.go [hunk 5]
```diff
@@ -105,6 +105,41 @@ func (s *Store) ListBuckets(accessKeyID string) ([]s3.BucketInfo, error) {
 	return buckets, err
 }
 
+// GetBucketVersioning returns the versioning status of the bucket. The status
+// is one of "" (never configured), "Enabled" or "Suspended".
+func (s *Store) GetBucketVersioning(accessKeyID, bucket string) (status string, err error) {
+	err = s.transaction(func(tx *txn) error {
+		bid, err := bucketID(tx, accessKeyID, bucket)
```
Comment (peterjan) [inline sia/persist/sqlite/buckets.go]: It feels like `bucketIDAndVersioning` would simplify things? I think I commented on that before but I'm not sure now 🤔 . I quickly checked all call sites and seems to make sense in all of them. 

While doing so I also noticed that `ExpireObjects` takes a bucket ID, which is not conventional. We expose the database ID in `AllBucketLifecycleConfigurations` with the note that says:

```
The bucket ID is used to apply rules so a bucket deleted and
// recreated under the same name mid-run cannot have stale rules applied to it;
// the name is informational.
```

Is this something that surfaced in the tox tests? I know it's unrelated to the PR but it feels like something we missed in the lifecycle PR

File: sia/persist/sqlite/versioning.go [hunk 6]
```diff
@@ -0,0 +1,238 @@
...
+	var id *sqlHash256
+	var sealed *sqlSiaObject
+	if siaObject != nil {
+		id = (*sqlHash256)(&siaObject.ID)
+		sealed = (*sqlSiaObject)(&siaObject.Sealed)
+	}
+
+	// the new row carries the highest seq and becomes the current version;
+	// clear the previous current version first to satisfy the unique index.
+	if err := clearLatest(tx, bid, name); err != nil {
```
Comment (peterjan) [inline sia/persist/sqlite/versioning.go]: `clearLatest` and `nextSeq` are always called one after the other, I think it would be cleaner to combine the two to make that more obvious, something like `claimCurrentSeq` perhaps?

File: sia/persist/sqlite/lifecycle.go [hunk 7]
```diff
@@ -146,25 +148,37 @@ func (s *Store) AbortMultipartUploads(bucketID int64, prefix string, before time
...
-// Deleted objects backed by Sia data are inserted into orphaned_objects; objects still
-// pending on disk have their newly-unreferenced filenames returned for cleanup.
-// It returns the number of objects deleted and performs no ownership checks.
-func (s *Store) ExpireObjects(bucketID int64, prefix string, before time.Time, limit int) (deleted int, orphans []sia.OrphanedFile, err error) {
+// ExpireObjects expires up to limit objects whose current version matches
+// prefix and was last modified at or before the cutoff, applying a
+// versioning-aware delete (see [deleteCurrentObject]) to each.
+// NoncurrentVersionExpiration is not supported. Returns the number of objects
+// expired; performs no ownership checks.
+func (s *Store) ExpireObjects(bucketID int64, prefix string, before time.Time, limit int) (deleted int, orphans []objects.OrphanedFile, err error) {
```
Comment (peterjan) [inline sia/persist/sqlite/lifecycle.go]: we updated ExpireObjects but did not update the accompanying test in lifecycle_test.go - probably worth extending it to make sure that it only applies to the current version

**Review by peterjan — CHANGES_REQUESTED** [review]
Tough PR to go through but I think it's fine. I would repeat the benchmark results you did in the previous PR here in the PR description for documenting purposes. Especially since you were able to make it so there's minimal regression.

File: sia/persist/sqlite/objects.go [hunk 8]
```diff
@@ -895,112 +1087,209 @@ func (s *Store) ListObjects(accessKeyID, bucket string, prefix s3.Prefix, page s
...
+// row's seq so paging can resume mid-key; a key-marker that lands inside a
+// common prefix is advanced past the whole group.
+func resolveVersionCursor(tx *txn, bid int64, prefix s3.Prefix, page s3.ListObjectVersionsPage) (versionMarker, error) {
+	var m versionMarker
+	if page.KeyMarker != nil {
+		m.key = *page.KeyMarker
+	}
+	if m.key != "" && page.VersionIDMarker != nil {
+		err := tx.QueryRow(`SELECT seq FROM objects WHERE bucket_id = $1 AND name = $2 AND version_id = $3`,
+			bid, m.key, *page.VersionIDMarker).Scan(&m.seq)
```
Comment (peterjan) [inline sia/persist/sqlite/objects.go]: *page.VersionIDMarker is the wire-encoded marker right? So it can be `null`? So this is probably a bug because we're feeding it straight into the query. if that's the case we're missing a test to surface this pagination bug

File: sia/persist/sqlite/buckets.go (hunk 5 above)
Comment (chris124567) [reply sia/persist/sqlite/buckets.go]: I agree it's a very unlikely edge case I changed it to take bucket name now

**Comment by peterjan** [comment]
hard to approve considering the build is failing but the PR looks good to me 👍

**Comment by chris124567** [comment]
Thank you
I'm going to request reviewers on #214 then

File: sia/persist/sqlite/objects.go [hunk 9]
```diff
@@ -895,112 +1057,213 @@ func (s *Store) ListObjects(accessKeyID, bucket string, prefix s3.Prefix, page s
...
+		versionID := *page.VersionIDMarker
+		if versionID == s3.Null {
+			versionID = nullVersion
+		}
+		err := tx.QueryRow(`SELECT seq FROM objects WHERE bucket_id = $1 AND name = $2 AND version_id = $3`,
+			bid, m.key, versionID).Scan(&m.seq)
+		if err == nil {
+			m.hasSeq = true
+		} else if !errors.Is(err, sql.ErrNoRows) {
+			return versionMarker{}, err
```
Comment (ChrisSchinnerl) [inline sia/persist/sqlite/objects.go:1134]: Isn't this an issue? If I wanted to resume listing mid-key but the version that the marker is currently pointed at was deleted in the meantime, would it continue with the next version or just skip over the whole object?

File: sia/persist/sqlite/objects_test.go
Comment (ChrisSchinnerl) [inline sia/persist/sqlite/objects_test.go:852]: Did you make sure this is still fast ?

**Review by ChrisSchinnerl — COMMENTED** [review]
Haven't been able to go through all of it in detail but it looks promising so far.

File: sia/persist/sqlite/objects_test.go
Comment (chris124567) [reply sia/persist/sqlite/objects_test.go:852]: Benchmark in PR description... minor slowdown

**Review by peterjan — DISMISSED** [review]

**Review by ChrisSchinnerl — CHANGES_REQUESTED** [review]
Looking good but got conflicts

**Review by ChrisSchinnerl — APPROVED** [review]
